nczarr_test/test_quantize.c

```diff
@@ -0,0 +1,935 @@
+/* This is part of the netCDF package.
+   Copyright 2021 University Corporation for Atmospheric Research/Unidata
+   See COPYRIGHT file for conditions of use.
+
+   Test quantization of netcdf-4 variables. Quantization is the
+   zeroing-out of bits in float or double data beyond a desired
+   precision.
+
+   Ed Hartnett, 8/19/21
+*/
+
+#include <nc_tests.h>
+#include "err_macros.h"
+#include "netcdf.h"
+
+#define DEBUG
+
+
+#define TEST "test_quantize"
+#define FILE_NAME "tmp_quantize"
+#define NDIM1 1
+#define DIM_NAME_1 "meters_along_canal"
+#define DIM_LEN_3 3
+#define DIM_LEN_1 1
+#define DIM_LEN_5 5
+#define DIM_LEN_8 8
+#define VAR_NAME_1 "Amsterdam_houseboat_location"
+#define VAR_NAME_2 "Amsterdam_street_noise_decibels"
+#define NSD_3 3
+#define NSD_9 9
+
+/* This var used to help print a float in hex. */
+char pf_str[20];
+
+/* This struct allows us to treat float as uint32_t
+ * types. */
+union FU {
+    float f;
+    uint32_t u;
+};
+
+/* This struct allows us to treat double points as uint64_t
+ * types. */
+union DU {
+    double d;
+    uint64_t u;
+};
+
+/* This function prints a float as hex. */
+char *
+pf(float myf)
+{
+    union {
+	float f;
+	uint32_t u;
+    } fu;
+    fu.f = myf;
+    sprintf(pf_str, "0x%x", fu.u);
+    return pf_str;
+}
+
+/* This function prints a double as hex. */
+char *
+pd(double myd)
+{
+    union {
+	double d;
+	uint64_t u;
+    } du;
+    du.d = myd;
+    sprintf(pf_str, "0x%lx", du.u);
+    return pf_str;
+}
+
+int
+main(int argc, char **argv)
+{
+    int retval = NC_NOERR;
+    const char* template = NULL;
+    char file_name[4096];
+    
+    if(argc == 1)
+        {fprintf(stderr,"usage: test_quantize <zarr-url-template>\n"); exit(1);}
+
+    template = argv[1];
+
+    snprintf(file_name,sizeof(file_name),template,FILE_NAME);
+#ifdef DEBUG
+    fprintf(stderr,"file_name=|%s|\n",file_name);
+#endif
+
+
+    printf("\n*** Testing netcdf-4 variable quantization functions.\n");
+    printf("**** testing quantization setting and error conditions...");
+    {
+        int ncid, dimid, varid1, varid2;
+        int quantize_mode_in, nsd_in;
+
+        /* Create a netcdf-4 file with two vars. Attempt
+         * quantization. It will work, eventually... */
+        if (nc_create(file_name, NC_NETCDF4|NC_CLOBBER, &ncid)) ERR;
```

Comment:
@DennisHeimbigner I don't understand how this tests the nczarr quantize. Seems like you would need to change the mode flags of the nc_create() calls, right? Otherwise these tests are just creating a netCDF/HDF5 file and testing quantize works there...

Reply:
The file_name is a URL in this case: e.g. s3://bucket/zarrfile#mode=nczarr,s3

Reply:
So you don't need to use NC_ZARR if the filename is a URL?

Does NC_NETCDF4 have any effect in this case?

Reply:
If the URL is properly structured, then even NC_NETCDF4 is not needed.